### DISCUSSION TITLE
Rodulf compatibility

### DISCUSSION THREAD

**Discussion by embcla** [gcormier/megadesk#118] [General]
Hi

Wonderful little hack!
I almost clicked buy, then I thought better check before.
My desk is a Rodulf rather than a Bekant.
Can you tell me what I should be looking for in order to verify whether Megadesk could work?
Is it an I2C protocol, do you have some messages that I could be looking for?

thanks!

**Comment by gcormier** [comment]
It's using LIN for communication. Saleae logic analyzer software can decode this, and the software works with cheap clones. I think pulseview/sigrok can as well.

**Comment by gnclmorais** [comment]
@embcla, did you end up looking into this? Is it compatible? I’m in the same position :)

**Comment by gcormier** [comment]
I can't even buy the RODULF anywhere near me. If someone is willing to open the controller, I'd be curious to see what it looks like inside.

**Comment by embcla** [comment]
Hi @gnclmorais no not yet, every evening I tell myself I should but am too tired to pull out the logic analyzer 😅 
Shall pull myself

**Comment by embcla** [comment]
Ok I went ahead and took the button module apart, no joy: it's purely analog
![20221129_100807](https://user-images.githubusercontent.com/54371771/204487520-ccf327d2-4ad9-4e7b-9a59-59668f2dc311.jpg)


I'll check with the controllers next

**Reply by embcla** [reply]
![20221129_105546](https://user-images.githubusercontent.com/54371771/204500140-cdea7cc2-dd4b-425a-8b74-d27c1fd3c6a4.jpg)


![20221129_110444](https://user-images.githubusercontent.com/54371771/204500001-b1c38479-6f72-4365-9497-605fccd7e420.jpg)

![20221129_110528](https://user-images.githubusercontent.com/54371771/204500004-a4010eff-927a-45ef-992f-8a615e14d78b.jpg)

First thing: don't go an touch this thing without discharging the capacitors!
Its auto-discharge is pretty low so even 15m after unplugging it, it's still able to give you a jolt.

So what you see is a single microcontroller on the logic, bunch of caps and resistors, and on each side, next to the 4-pin molex connector to the motor driver, 2 smaller chips labeled H004 CE RB16. Couldn't find anything, but they have the look of power chips, so either stepper drivers or quarter bridges.

I suspect this is a very different project from the Bekant: the layout hints at having some half bridge setup on each driver connector. I was hoping the fact each driver had 4 pins meant there was some logic going, but now I think it's just the end stops plus power.

There's the chip programming header there, but no other unpopulated footprint apart from a capacitor. So I doubt there is any ability to add functionality. The chip is most probably not listening for anything. 

The other side of this, is that it look very easy to do anything you want with this: redoing the entire software to autonomously listen for button clicks and do memories seems possible.

It needs a more detailed analysis to understand whether there are free GPIOs to add some feedback (buzzer/light)

It could be possible to lift the program if it's not read protected, but I'm not able to decode machine code, I'm a pure C guy, anything below that escapes me 😆 
If anyone can suggest easier way to read unknown code (is there anything out there that can reconstruct lifted code?) I'd happily connect to it with a debugger and try lifting the program

**Reply by gcormier** [reply]
If it's not protected (which I'd wager, it is protected), it would be assembly code that you'd get.


![image](https://user-images.githubusercontent.com/8388983/204865813-7fc62941-0ebf-4791-a007-f11f02753644.png)

The TX/RX on the right side, if that is where the up/down button cable is plugged in (ethernet style), it might double as a programming port. Otherwise the buttons are quite dumb.

So that leaves us with the half(?) bridges for the motors potentially? For the 4 pin header, 2 pins are high power, and 2 pins are low power based on the tiny traces?. A few thoughts

- The 2 low power pins would likely be ground and signal - perhaps still LIN? If you probe the points, once you identify ground, if the other 2 pins are a short-circuit then it's probably a LINbus again. They could also be isolated busses, either LIN or just a homebrew 1-wire protocol.
- Unsure why they wouldn't just provide high current +/- to the motors and why the MOSFET's are needed. The motors will need power to provide digital status to the controller.
- Maybe the motors are dumb - calibration simply has them go down as far as they can until they can't anymore and they are current-limited, after which as long as steps are counted and not missed, the controller just assumes the motors have properly moved the correct number of steps - comparable to a 3d printer, after homing it can finely control steppers with no feedback whatsoever, and if a step is lost, the print is affected. In which case all 4 leads could be high powered?

**Reply by macmoritz** [reply]
The PIC16F18857 is LIN combatible (found in the datasheet). Therefore it could be that the motors are controlled via LIN.

I have no knowledge about LIN, but this is a bus protocol, so we could jump into the system anywhere, or am I wrong?

**Comment by embcla** [comment]
I can't figure out the motor sections without finding out what chips there are what circuit they make so I turned to look elsewhere.

The control board has only 2 buttons and 3 wires, the other 5 in the plug are not populated, but I went looking at the pcb footprint and there is something interesting

![20221201_120809](https://user-images.githubusercontent.com/54371771/205038357-b573fa06-4ec9-45dd-a477-57e74f9f70e9.jpg)
There are labels for TX and RX.

Is it the same for the Bekant? Do you have a LIN-TX and LIN-RX?

**Reply by gcormier** [reply]
No - Bekant is LIN which is a 1 wire protocol.

**Reply by macmoritz** [reply]
As I understand LIN is based on UART. Maybe LIN uses one of the pins used for UART?
Maybe the[ information below](https://github.com/gcormier/megadesk/discussions/118#discussioncomment-5432442) to the voltage levels of TX and RX pin helps.

**Comment by thetom88** [comment]
do you guys have any update on this topic? I'd be so happy to control my Rudolf table with google assistant or alexa one day.. My stupid idea was to intercept the buttons and use an ultrasonic sensor to measure the table height, but then I came across this project.

**Reply by gcormier** [reply]
I'd need better pictures of both sides of the board to even begin guessing how it works, but it certainly much different then the BEKANT.

**Comment by macmoritz** [comment]
Hi there
I have a Rodulf desk and want to share my findings until here and maybe get some help or ideas from you.

Firstly I started figuring out the pinout of the RJ-45 connector for the controller. Rather easy, thanks to @embcla! Note that there is one pin, between TX and UP not used.

- as expected pulling DN and UP to GND moves the desk.
- RST does what we can think, it performs a reset. This causes the need to move the desk to the lowest position
- the RX pin carries an voltage of about 0V, TX carries around 5V.


I hooked RX and TX up to an arduino uno with softwareserial but no data gets received from the serial communication.
Do you have any ideas to try to get a connection running?

**Reply by gcormier** [reply]
RX/TX sound like just dumb lines, logic level, that are either on/off based on the button presses.

The actual communication to the motors will be via the motor connector pins. However, since there's large MOSFEts, I think two of these pins will be higher voltage. It could be that one of the pins is encoder feedback or a LIN bus.

**Reply by macmoritz** [reply]
Sadly TX/RX does not change at button press.

I guess that the Pins are for programming or they need some kind of signal to start the communication.

**Reply by gcormier** [reply]
Oh sorry, I mixed up the UP/DN with RX/TX.

You could buy a cheap logic analyzer to see if either RX/TX are changing. Something like https://www.amazon.ca/Comidox-Analyzer-Device-Channel-Arduino/dp/B07KW445DJ , available on AliExpress as well.

**Comment by H1D** [comment]
Just a note. RODULF does actually has a memory feature. Here is how to set it up:
1) set height to desired "low" position, then press **both button 3 times**, then press "down" button. This way you set you lowest position
2) do the same for upper position, but use "up" button after pressing both 3 times.
3) now the table moves betwen those two positions only, but you have to **hold** a button to do so.
4) to "break throught" min/max position you'll have to hold up/down button for approx 20 seconds.

So its just a matter of shorting 2 pins for about 10 seconds. It can be easily done with at esphome or something. Also simple 2-way switch would to the trick (but you'll have to switch it to neutral in 10 secs)

**Reply by CM000n** [reply]
There was also a discussion about bridging the connections in the Home Assistant forum with instructions: https://github.com/jensihnow/smarthome/blob/main/smartdesk/README.md
Of course, it would be cool if there was something like this out of the box for the lazy tinkerers 😄

**Comment by embcla** [comment]
@H1D oh my god, thank you so very much! <3
I just tried it, it works!
How and where did you find this?

**Reply by gcormier** [reply]
Upsy Desky might work if it's RJ45

**Reply by H1D** [reply]
> How and where did you find this?

some random reddit

**Reply by Zut2020** [reply]
Its also in the manual.
